plugins/cilium-cni/cilium-cni.go

```diff
@@ -389,25 +390,74 @@ func cmdAdd(args *skel.CmdArgs) (err error) {
-	if err != nil {
-		err = fmt.Errorf("unable to allocate IP via local cilium agent: %s", err)
-		return
-	}
+	if conf.IpamMode == ipamOption.IPAMDelegatedPlugin {
+		// Skip the call to cilium-agent IPAMAllocate and use the delegated IPAM plugin instead.
+		var ipamRawResult cniTypes.Result
+		var ipamResult *cniTypesVer.Result
+		ipamRawResult, err = cniInvoke.DelegateAdd(context.TODO(), n.IPAM.Type, args.StdinData, nil)
+		if err != nil {
+			err = fmt.Errorf("delegated IPAM cmd ADD failed: %s", err)
+			return
+		}
 
-	if ipam.Address == nil {
-		err = fmt.Errorf("Invalid IPAM response, missing addressing")
-		return
-	}
+		// CNI spec says if an error occurs, we should call the delegated IPAM plugin
+		// again with a DEL command so it can release the IPs.
+		defer func() {
+			if err != nil {
+				cniInvoke.DelegateDel(context.TODO(), n.IPAM.Type, args.StdinData, nil)
+			}
+		}()
+
+		ipamResult, err = cniTypesVer.NewResultFromResult(ipamRawResult)
+		if err != nil {
+			err = fmt.Errorf("could not interpret delegated IPAM result for CNI version %s: %s", cniTypesVer.ImplementedSpecVersion, err)
+			return
+		}
+
+		// The rest of the CNI code uses models.IPAMResponse, so convert the output of the
+		// IPAM plugin to that format. TODO: decide whether this is a reasonable approach.
+		ipam = &models.IPAMResponse{
+			HostAddressing: conf.Addressing,
+			Address:        &models.AddressPair{},
+		}
+
+		// Safe to assume at most one IP per family. The K8s API docs say:
+		// "Pods may be allocated at most 1 value for each of IPv4 and IPv6"
+		// https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-v1/
+		for _, ipConfig := range ipamResult.IPs {
+			// TODO: there are two fields in ipamResult that we're ignoring here: Routes and Gateways.
+			// There doesn't seem to be a way to inject those into models.IPAMResponse, but maybe
+			// that's something we'd need to add?
+			ipNet := ipConfig.Address
+			if ipv4 := ipNet.IP.To4(); ipv4 != nil {
+				ipam.Address.IPV4 = ipNet.String()
+				ipam.IPV4 = &models.IPAMAddressResponse{IP: ipv4.String()}
+			} else {
+				ipam.Address.IPV6 = ipNet.String()
+				ipam.IPV6 = &models.IPAMAddressResponse{IP: ipNet.IP.String()}
+			}
+		}
+	} else {
+		podName := string(cniArgs.K8S_POD_NAMESPACE) + "/" + string(cniArgs.K8S_POD_NAME)
+		ipam, err = c.IPAMAllocate("", podName, true)
+		if err != nil {
+			err = fmt.Errorf("unable to allocate IP via local cilium agent: %s", err)
+			return
+		}
 
-	// release addresses on failure
-	defer func() {
-		if err != nil && ipam != nil && ipam.Address != nil {
-			releaseIP(c, ipam.Address.IPV4)
-			releaseIP(c, ipam.Address.IPV6)
+		if ipam.Address == nil {
+			err = fmt.Errorf("Invalid IPAM response, missing addressing")
+			return
 		}
-	}()
+
+		// release addresses on failure
+		defer func() {
+			if err != nil && ipam != nil && ipam.Address != nil {
+				releaseIP(c, ipam.Address.IPV4)
+				releaseIP(c, ipam.Address.IPV6)
+			}
+		}()
```

Comment:
We still need to release any allocated IPs on later failure for the delegated plugin case as well, right?

Reply:
there's a `defer` on line 405 above to invoke the delegated plugin's DEL, which should release the IP. I was thinking we'd skip the call to cilium-daemon's `IPAMReleaseIP`, but maybe there's some cleanup there we'd still need?

Reply:
Ah right, I just missed this. No, as far as I know this should be good enough.